File: 2023/P06/problem6a.rs
use std::fs;

fn main() {
    let data = fs::read_to_string("problem6.txt")
        .expect("Failed to read file");
    let data = data
        .trim()
        .split('\n')
        .collect::<Vec<_>>();
    let times = data[0]
        .split(": ")
        .collect::<Vec<_>>()[1]
        .split(" ")
        .filter(|&x| x != "")
        .map(|x| x
            .parse::<i32>()
            .expect("Failed to parse")
        )
        .collect::<Vec<_>>();
    let distances = data[1]
        .split(": ")
        .collect::<Vec<_>>()[1]
        .split(" ")
        .filter(|&x| x != "")
        .map(|x| x
            .parse::<i32>()
            .expect("Failed to parse")
        )
        .collect::<Vec<_>>();
    let mut prod = 1;
    for (t, d) in times.iter().zip(distances.iter()) {
        let mid: f64 = *t as f64;
        let disc: f64 = ((t * t - 4 * d) as f64).sqrt();
        let lo = (mid - disc) / 2.0;
        let hi = (mid + disc) / 2.0;
        let lo: u32 = lo.ceil() as u32;
        let hi: u32 = hi.floor() as u32;
        prod *= hi - lo + 1;
    }
    println!("{prod}");
}

File: 2024/P06/problem6a.py
data = []
with open("problem6.txt", 'r') as f:
    data = f.read().strip('\n').split('\n')

guard_position = None
for i in range(len(data)):
    for j in range(len(data[i])):
        if data[i][j] == '^':
            guard_position = (i, j)

positions = set()
direc = (-1, 0)
while True:
    x, y = guard_position
    if x < 0 or x >= len(data) or y < 0 or y >= len(data[x]):
        break
    positions.add(guard_position)
    dx, dy = direc
    nx, ny = (x + dx, y + dy)
    while 0 <= nx < len(data) and 0 <= ny < len(data[nx]) and data[nx][ny] == '#':
        dx, dy = (dy, -dx)
        nx, ny = (x + dx, y + dy)
    guard_position = (nx, ny)
    direc = dx, dy

print(len(positions))
